File: java-patterns/src/main/java/behavioral/template/Template.java
package behavioral.template;

/**
 * Usage: 考卷 <br/>
 * Date: 2023/4/28 11:11 <br/>
 *
 * @author <a href="mailto:hanzhang2566@foxmail.com">hanzhang</a>
 */
public abstract class Template {
    protected void template() {
        System.out.println("who are u?");
        System.out.println("I'm " + whoami());
    }

    /**
     * {@link Template#template()} 的答案
     *
     * @return 答案
     */
    protected abstract String whoami();
}

File: go-patterns/behavioral/template/template.go
package template

import "fmt"

type templateIfc interface {
	whoami() string
}

type Template struct {
	steps templateIfc
}

func (t Template) temp() {
	fmt.Println("who are u?")
	fmt.Printf("I'm %s\n", t.whoami())
}

func (t Template) whoami() string {
	return t.steps.whoami()
}

type pig struct {
}

func (p pig) whoami() string {
	return "pig"
}
